pkg/handler/iam_handler.go

```diff
@@ -0,0 +1,201 @@
+			return nil, fmt.Errorf("failed to apply client options: %w", err)
+		}
+	}
+	h.organizationsClient = organizationsClient
+	h.foldersClient = foldersClient
+	h.projectsClient = projectsClient
+
+	if h.retry == nil {
+		h.retry = retry.WithMaxRetries(5, retry.NewFibonacci(500*time.Millisecond))
+	}
+	return h, nil
+}
+
+// Do removes expired or duplicative IAM bindings added by AOD and adds requested IAM bindings to current IAM policy.
+func (h *IAMHandler) Do(ctx context.Context, r *v1alpha1.IAMRequestWrapper) (nps []*v1alpha1.IAMResponse, retErr error) {
+	for _, p := range r.ResourcePolicies {
+		np, err := h.handlePolicy(ctx, p, r.Duration)
+		if err != nil {
+			retErr = errors.Join(
+				retErr,
+				fmt.Errorf("failed to handle policy update for resource %s: %w", p.Resource, err),
+			)
+		}
+		if np != nil {
+			nps = append(nps, np)
+		}
+	}
+	return
+}
+
+func (h *IAMHandler) handlePolicy(ctx context.Context, p *v1alpha1.ResourcePolicy, ttl time.Duration) (*v1alpha1.IAMResponse, error) {
+	var iamC IAMClient
+	switch strings.Split(p.Resource, "/")[0] {
+	case "organizations":
+		iamC = h.organizationsClient
+	case "folders":
+		iamC = h.foldersClient
+	case "projects":
+		iamC = h.projectsClient
+	default:
+		return nil, fmt.Errorf("resource isn't one of [organizations, folders, projects]")
+	}
+
+	var np *iampb.Policy
+	if err := retry.Do(ctx, h.retry, func(ctx context.Context) error {
+		// Get current IAM policy.
+		cp, err := iamC.GetIamPolicy(ctx, &iampb.GetIamPolicyRequest{Resource: p.Resource})
+		if err != nil {
+			return fmt.Errorf("failed to get IAM policy: %w", err)
+		}
+
+		// Update the policy with new IAM binding additions.
+		updatePolicy(cp, p.Bindings, ttl)
+
+		// Set the new policy.
+		setIamPolicyRequest := &iampb.SetIamPolicyRequest{
+			Resource: p.Resource,
+			Policy:   cp,
+		}
+		np, err = iamC.SetIamPolicy(ctx, setIamPolicyRequest)
+		// Retry when set IAM policy fail.
+		// TODO(#8): Look for specific errors to retry.
+		if err != nil {
+			return retry.RetryableError(fmt.Errorf("failed to set IAM policy: %w, retrying", err))
+		}
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to handle IAM request: %w", err)
+	}
+
+	return &v1alpha1.IAMResponse{Resource: p.Resource, Policy: np}, nil
+}
+
+// Remove expired bindings and add or update new bindings with expiration condition.
+func updatePolicy(p *iampb.Policy, bs []*v1alpha1.Binding, ttl time.Duration) {
+	// Convert new bindings to a role to unique bindings map.
+	bsMap := toBindingsMap(bs)
+	// Clean up current policy bindings.
+	var result []*iampb.Binding
+	for _, cb := range p.Bindings {
+		// Skip non-AOD bindings.
+		if cb.Condition == nil || cb.Condition.Title != ConditionTitle {
+			result = append(result, cb)
+			continue
+		}
+		// TODO (#6): Remove expired bindings.
+		// Exclude duplicative Members from current bindings.
+		var nm []string
+		for _, m := range cb.Members {
+			if _, ok := bsMap[cb.Role][m]; !ok {
```

Comment:
There is no guarantee `bsMap[cb.Role]` is in the map in the first place? Should we check it the first thing in the for loop?

Reply:
Done.